lib/src/component/with_transition.dart

```diff
@@ -0,0 +1,410 @@
+///         (Dom.div()
+///           ..className = classes.toClassName()
+///         )(
+///           // Some children...
+///         ),
+///       )
+///     );
+///   },
+///   $WithTransitionExampleConfig, // ignore: undefined_identifier
+/// );
+/// ```
+///
+/// __Wrapping around custom components__
+///
+/// If you want to wrap around another component instead of around a single `Dom` element child as shown in the
+/// first example above, you must ensure that the value of `props.ref` is forwarded to the `Element` that has
+/// the CSS transition using `uiForwardRef`:
+///
+/// ```dart
+/// import 'package:over_react/over_react.dart';
+///
+/// mixin CustomChildProps on UiProps {}
+///
+/// UiFactory<CustomChildProps> CustomChild = uiForwardRef(
+///   (props, ref) {
+///     return (Dom.div()..ref = ref)(
+///       props.children,
+///     ),
+///   },
+///   $CustomChildConfig, // ignore: undefined_identifier
+/// );
+/// ```
+UiFactory<WithTransitionProps> WithTransition = memo(uiFunction(
+  (_props) {
+    // ----- Default Props ----- //
+    final defaultProps = useMemo(() {
+      return WithTransition()
+      ..isShown = false
+      ..transitionTimeout = const Duration(seconds: 1)
+      ..childPropsByPhase = const {
+        v2.TransitionPhase.PRE_SHOWING: {},
+        v2.TransitionPhase.SHOWING: {},
+        v2.TransitionPhase.SHOWN: {},
+        v2.TransitionPhase.HIDING: {},
+        v2.TransitionPhase.HIDDEN: {},
+      }
+      ..addProps(v2.TransitionPropsMixin.defaultProps);
+    }, const []);
+    final props = defaultProps..addProps(_props);
+
+    assert(_hasSingleValidChild(props));
+
+    // ----- State / Hooks ----- //
+    final transitionNodeRef = useRef<Element>();
+    final transitionPhaseDerivedFromProps = props.isShown ? TransitionPhase.SHOWN : TransitionPhase.HIDDEN;
+    final transitionPhase = useState(transitionPhaseDerivedFromProps);
+
+    /// Begin showing the node unless the [currentPhase] is already shown or is in the process of showing.
+    void _handleShow(WithTransitionProps tProps, TransitionPhase currentPhase) {
+      if (_isOrWillBeShown(currentPhase)) return;
+
+      tProps.onWillShow?.call();
+      transitionPhase.set(_hasTransitionIn(tProps) ? TransitionPhase.PRE_SHOWING : TransitionPhase.SHOWN);
+    }
+
+    /// Begin hiding node unless the [currentPhase] is already hidden or is in the process of hiding.
+    void _handleHide(WithTransitionProps tProps, TransitionPhase currentPhase) {
+      if (_isOrWillBeHidden(currentPhase)) return;
+
+      tProps.onWillHide?.call();
+      transitionPhase.set(
+          _hasTransitionOut(tProps) ? TransitionPhase.HIDING : TransitionPhase.HIDDEN);
+    }
+
+    // getDerivedStateFromProps
+    if (props.isShown && _isOrWillBeHidden(transitionPhase.value)) {
+      _handleShow(props, transitionPhase.value);
+    } else if (!props.isShown && _isOrWillBeShown(transitionPhase.value)) {
+      _handleHide(props, transitionPhase.value);
+    }
```

Comment:
These calls mutating state and performing side-effects in `render`, which isn't allowed; they should go in `useEffect`

Reply:
Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8